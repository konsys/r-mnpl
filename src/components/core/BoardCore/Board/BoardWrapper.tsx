import "./style.scss";
import "react-toastify/dist/ReactToastify.css";
import "../../../../theme/styles/board/theme.scss";

import { BoardMessage, FieldStatus, IPlayer } from "types/types";
import { InitBoardPlayersGate, players$ } from "stores/Board/PlayersStore";
import React, { useEffect } from "react";
import {
  actions$,
  setCurrentActionEvent,
} from "../../../../stores/Board/ActionStore";
import {
  fieldAction$,
  fields$,
  setFieldsEvent,
} from "stores/Board/FieldsStore";
import { useGate, useStore } from "effector-react";

import { Arbitr } from "../../../views/BoardViews/Arbitr/Arbitr";
import { BoardDomain } from "stores/Board/BoardDomain";
import { BoardModal } from "../../../views/BoardViews/BoardModal/BoardModal";
import { BoardSocketGate } from "stores/Board/SocketConnetcModel";
import { BoardView } from "components/views/BoardViews/Board/BoardView";
import { Chat } from "../../../views/BoardViews/Chat/Chat";
import { Contract } from "../../../views/BoardViews/Contract/Contract";
import { Dices } from "../../../views/BoardViews/Dices/Dices";
import { IRoomState } from "stores/Game/Rooms/RoomsModel";
import { M1tv } from "../../../views/BoardViews/M1tv/M1tv";
import { Players } from "components/views/BoardViews/Players/Players";
import { TableHelper } from "../../../views/BoardViews/TableHelper/TableHelper";
import { Ticket } from "../../../views/BoardViews/Ticket/Ticket";
import { ToastContainer } from "react-toastify";
import { Tokens } from "../../../views/BoardViews/Tokens/Tokens";
import { clearNode } from "effector";
import { contract$ } from "stores/Board/ContractStore";
import { isEqual } from "lodash";
import nanoid from "nanoid";
import { tokens$ } from "stores/Board/TokensStore";
import { updateAllPlayers } from "utils/players.utils";
import { user$ } from "stores/Game/User/UserModel";

export const fieldsHandler = (messageFieldsStatus: FieldStatus[]) => {
  allFieldsIterate(messageFieldsStatus);
};

// Not used
export const statusFieldsIterate = (messageFieldsStatus: FieldStatus[]) => {
  const store = fields$.getState();
  let toUpdateStore = false;

  messageFieldsStatus.forEach((status) => {
    const storeFieldIndex = store.fields.findIndex(
      (storeField) => storeField.fieldId === status.fieldId
    );

    const currentField = store.fields[storeFieldIndex];

    if (currentField && !isEqual(status, currentField.status)) {
      store.fields[storeFieldIndex] = {
        ...store.fields[storeFieldIndex],
        status,
      };
      toUpdateStore = true;
    }
  });

  toUpdateStore && setFieldsEvent({ ...store, version: ++store.version });
};

const allFieldsIterate = (messageFieldsStatus: FieldStatus[]) => {
  const store = fields$.getState();

  store.fields &&
    store.fields.forEach((storeField, index) => {
      const messageFieldStatus = messageFieldsStatus.find(
        (messageField) => messageField.fieldId === storeField.fieldId
      );

      if (messageFieldStatus) {
        store.fields[index] = {
          ...store.fields[index],
          status: messageFieldStatus,
        };
      } else {
        store.fields[index] = {
          ...store.fields[index],
          status: undefined,
        };
      }
    });
  setFieldsEvent({ ...store, version: ++store.version });
};

export const MessageHandler = (message: BoardMessage) => {
  const event = message.data.event.action;
  fieldsHandler(message.data.boardStatus.fields);
  playersHandler(message.data.boardStatus.players);

  setCurrentActionEvent({
    actionId: (event && event._id) || nanoid(4),
    event: message.data.event,
  });
};
export const playersHandler = (players: IPlayer[]) => updateAllPlayers(players);

export const BoardWrapper = ({ board }: { board: IRoomState }) => {
  const playerIds = board ? board.players.map((v) => (v ? v.userId : -1)) : [];

  useGate(InitBoardPlayersGate, { userIds: playerIds, user: "me" });
  useGate(BoardSocketGate, "");

  useEffect(() => {
    return () => {
      clearNode(BoardDomain, { deep: true });
    };
  }, []);

  const action = useStore(actions$);
  const playersData = useStore(players$);
  const { fields } = useStore(fields$);
  const fieldActionId = useStore(fieldAction$);
  const contract = useStore(contract$);
  const user = useStore(user$);
  const tokensData = useStore(tokens$);
  const tokens = tokensData && tokensData.tokens ? tokensData.tokens : [];

  const players =
    playersData && playersData.players && playersData.players.length
      ? playersData.players
      : null;

  const modal = () => {
    try {
      return (
        user &&
        action &&
        user.userId === action.event.action.userId && (
          <BoardModal
            isModal={
              (action.event.action && action.event.action.isModal) || false
            }
          />
        )
      );
    } catch (err) {}
  };

  return (
    <>
      {players &&
        players.length &&
        fields.length &&
        tokens.length &&
        user &&
        user.userId && (
          <div
            className="boardWrapper"
            style={{ width: "100%", height: "100%" }}
            mnpl-newcolors="1"
            cz-shortcut-listen="true"
            mnpl-mouse="1"
            data-os="null"
            data-browser="chrome"
          >
            <div className="table _shakehack">
              <div className="table-body">
                <Players players={players} user={user} action={action} />
                <div className="table-body-board">
                  <BoardView
                    fields={Array.isArray(fields) ? fields : []}
                    fieldActionId={fieldActionId}
                    contract={contract}
                    user={user}
                    players={players}
                  />

                  <div className="table-body-board-center">
                    <M1tv />
                    {modal()}
                    <Arbitr />
                    <Ticket />
                    <Chat />
                  </div>
                  <div className="table-body-board-tokens">
                    <Tokens tokens={tokens} fields={fields} />
                  </div>
                  <Dices />
                  <Contract user={user} contract={contract} />
                  <TableHelper />
                </div>
              </div>

              <div id="placeholder_gameover" />
            </div>
            <div className="table-jokes" />
          </div>
        )}

      <ToastContainer />
    </>
  );
};
