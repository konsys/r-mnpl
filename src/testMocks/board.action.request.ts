import { IBoardActionRequest, OutcomeMessageType } from "types/types";

export const testBoardActionRequest:IBoardActionRequest = {
    gameId:'testGameId',
    action:OutcomeMessageType.OUTCOME_ROLL_DICES_CLICKED
}