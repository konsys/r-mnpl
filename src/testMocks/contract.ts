import { IContract } from "types/types";

export const testContract: IContract = {
  fieldFromPrice: 1000,
  fieldIdsFrom: [1, 2],
  fieldIdsTo: [3],
  fieldToPrice: 1000,
  fromUserId: 1,
  moneyFrom: 1000,
  moneyTo: 1000,
  toUserId: 2,
};
