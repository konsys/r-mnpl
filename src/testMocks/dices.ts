import { IDices, IncomeMessageType } from "types/types";

export const testDices: IDices = {
    _id: "testId",
    isModal: true,
    type: IncomeMessageType.INCOME_ROLL_DICES_MODAL,
    dices: [1, 2, 3],
    dicesSum: 6,
    userId: 1,
  };
  