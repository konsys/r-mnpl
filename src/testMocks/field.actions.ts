import { testField, testMortgagedField, testOwnedField } from "./field";

import { IFieldModal } from "components/views/BoardViews/Field/FieldActions/FieldActions";

export const testFieldActions: IFieldModal = {
  ...testField,
  isActive: true,

  position: { top: 20, left: 20 },
};

export const testOwnedFieldActions: IFieldModal = {
  ...testOwnedField,
  isActive: true,

  position: { top: 20, left: 20 },
};

export const testStaredFieldActions: IFieldModal = {
  ...testOwnedField,
  isActive: true,

  position: { top: 20, left: 20 },
};

export const testMortgagedFieldActions: IFieldModal = {
  ...testMortgagedField,
  isActive: true,

  position: { top: 20, left: 20 },
};
