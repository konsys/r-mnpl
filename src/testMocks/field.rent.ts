import { IFieldRent } from "types/types";

export const testRent: IFieldRent = {
  baseRent: 100,
  monopolyRent: 150,
  oneStar: 300,
  twoStar: 900,
  freeStar: 2000,
  fourStar: 2500,
  bigStar: 4000,
  paymentMultiplier: 2,
};
