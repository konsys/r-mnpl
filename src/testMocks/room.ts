import {
  IRoomResponce,
  IRoom,
  PlayerRoomStatus,
  RoomPortalFieldType,
  RoomStatus,
  RoomType,
} from "stores/Game/Rooms/RoomsModel";
import { testPlayer1, testPlayer2 } from "./user";

export const testPlayingRoom: IRoom = {
  roomId: "testRoomIdtestPlayingRoom",
  creatorId: 1,
  winnerId: null,
  players: [
    { ...testPlayer1, playerRoomStatus: PlayerRoomStatus.ACITVE },
    { ...testPlayer2, playerRoomStatus: PlayerRoomStatus.ACITVE },
  ],
  createTime: new Date(),
  roomType: RoomType.REGULAR,
  playersNumber: 2,
  autostart: false,
  privateRoom: false,
  restarts: false,
  portalType: RoomPortalFieldType.NOP,
  roomStatus: RoomStatus.PLAYING,
};

export const test4Room: IRoom = {
  roomId: "testRoomIdtest4Room",
  creatorId: 1,
  winnerId: null,
  players: [
    { ...testPlayer1, playerRoomStatus: PlayerRoomStatus.ACITVE },
    { ...testPlayer2, playerRoomStatus: PlayerRoomStatus.ACITVE },
  ],
  createTime: new Date(),
  roomType: RoomType.REGULAR,
  playersNumber: 4,
  autostart: false,
  privateRoom: false,
  restarts: false,
  portalType: RoomPortalFieldType.NOP,
  roomStatus: RoomStatus.PLAYING,
};

export const testPendingRoom1: IRoom = {
  roomId: "testRoomIdtestPendingRoom1",
  creatorId: 1,
  winnerId: null,
  players: [
    { ...testPlayer1, playerRoomStatus: PlayerRoomStatus.ACITVE },
    { ...testPlayer2, playerRoomStatus: PlayerRoomStatus.ACITVE },
  ],
  createTime: new Date(),
  roomType: RoomType.REGULAR,
  playersNumber: 2,
  autostart: false,
  privateRoom: false,
  restarts: false,
  portalType: RoomPortalFieldType.NOP,
  roomStatus: RoomStatus.PENDING,
};

export const testPendingRoom2: IRoom = {
  roomId: "testRoomIdtestPendingRoom2",
  creatorId: 2,
  winnerId: null,
  players: [
    { ...testPlayer1, playerRoomStatus: PlayerRoomStatus.ACITVE },
    { ...testPlayer2, playerRoomStatus: PlayerRoomStatus.ACITVE },
  ],
  createTime: new Date(),
  roomType: RoomType.REGULAR,
  playersNumber: 2,
  autostart: false,
  privateRoom: false,
  restarts: false,
  portalType: RoomPortalFieldType.NOP,
  roomStatus: RoomStatus.PENDING,
};

export const testRoomsResponce: IRoomResponce = {
  playersInRooms: 2,
  rooms: [testPlayingRoom],
};

export const testPendingRoomsResponce: IRoomResponce = {
  playersInRooms: 2,
  rooms: [testPendingRoom1, testPendingRoom2],
};
