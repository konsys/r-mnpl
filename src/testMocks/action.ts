import { ICurrentAction } from "stores/Board/ActionStore";
import {
  BoardAction,
  IncomeMessageType,
  OutcomeMessageType,
} from "types/types";

export const testDoNothingAction: ICurrentAction = {
  actionId: "testActionId",
  event: {
    action: {
      _id: "testID",
      isModal: false,
      userId: 1,
      type: IncomeMessageType.DO_NOTHING,
    },
  },
};

export const testRollDicesModal: BoardAction = {
  _id: "testBoardAction",
  type: IncomeMessageType.INCOME_ROLL_DICES_MODAL,

  userId: 1,
  title: "testTitleRollDices",
  text: "testTextRollDices",
  isModal: true,
};
export const testRollDicesClicked: BoardAction = {
  _id: "testBoardAction",
  type: OutcomeMessageType.OUTCOME_ROLL_DICES_CLICKED,
  userId: 1,
  isModal: false,
};
