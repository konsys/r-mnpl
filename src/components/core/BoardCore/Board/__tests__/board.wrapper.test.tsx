import { BoardWrapper } from "../BoardWrapper";
import React from "react";
import { shallow } from "enzyme";
import { testRoom } from "testMocks/room";

describe("Board wrapper test", () => {
  it("should renders", () => {
    expect(shallow(<BoardWrapper board={testRoom} />)).toMatchSnapshot();
  });
});
