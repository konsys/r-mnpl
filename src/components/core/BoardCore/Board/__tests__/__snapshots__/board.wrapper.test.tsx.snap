// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board wrapper test should renders 1`] = `ShallowWrapper {}`;
